### PR TITLE
hparams: add Python layer for hparams and metrics

### PR DESCRIPTION
**Pull request by wchargin**
Summary:
This change introduces `HParam`, `Metric`, and `Experiment` classes,
which represent their proto counterparts in a more Python-friendly way.
It similarly includes a `Domain` class hierarchy, which does not
correspond to a specific proto message, but rather unifies the domain
variants defined on the `HParamInfo` proto.

The design is roughly as in the original sketch of #1998.

The primary benefit of this change is that having first-class domains
enables clients to reuse the domain information for both the experiment
summary and the underlying tuning algorithm. We don’t provide a method
to do this out of the box, because we don’t actually provide any tuners
at this time, but it’s easy to write (e.g.) a `sample_uniform` function
like the one included in this commit. Then, sampling is as easy as

```python
    hparams = {h: sample_uniform(h.domain, rng) for h in HPARAMS}
```

It is also now more convenient to reference hparam values such that
static analysis can detect potential typos, because the `HParam` objects
themselves can be declared as constants and used as keys in a dict.
Writing `hparams["dropuot"]` fails at runtime, but `hparams[HP_DROPUOT]`
fails at lint time.

As a pleasant bonus, hparam definitions are now more compact, fitting
on one line instead of several. The demo code has net fewer lines.

Manual summary writer management is still required. A future change will
introduce a Keras callback to reduce this overhead.

Test Plan:
Some unit tests included, and the demo still works.

wchargin-branch: hparams-structured-api

### CODE REVIEW

**Comment by wchargin** [comment]
cc @erzel

**Comment by wchargin** [comment]
Reassigning to @stephanwlee to distribute load. For the other two in
particular I’d like @nfelt’s feedback on the summary APIs, but this one
is mostly summary-agnostic.

File: tensorboard/plugins/hparams/BUILD [hunk 1]
```diff
@@ -146,6 +146,38 @@ sh_test(
...
+        "@org_pythonhosted_six",
+    ],
+)
+
+py_test(
+    name = "api_test",
+    srcs = ["api_test.py"],
+    srcs_version = "PY2AND3",
+    visibility = [
+        "//visibility:public",
```
Comment (stephanwlee) [inline tensorboard/plugins/hparams/BUILD]: probably don't need to be really public, right?

File: tensorboard/plugins/hparams/api.py [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+
+from tensorboard.plugins.hparams import api_pb2
+from tensorboard.plugins.hparams import summary
+
+
+class Experiment(object):
+  """A top-level experiment description.
+
+  An experiment has a fixed set of hyperparameters and metrics, and
+  consists of multiple sessions. Each session has different associated
```
Comment (stephanwlee) [inline tensorboard/plugins/hparams/api.py]: Here, is it okay if I interpret `session` as `tf.Session`? Or does it map better to TB's `run`s?

File: tensorboard/plugins/hparams/api.py [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+    if dtype not in (int, float, bool, str):
+      raise ValueError("Unknown dtype: %r" % (dtype,))
+    self._dtype = dtype
+    for value in self._values:
+      if not isinstance(value, self._dtype):
+        raise ValueError(
+            "dtype mismatch: not isinstance(%r, %s)"
+            % (value, self._dtype.__name__)
+        )
+    self._values.sort()
```
Comment (stephanwlee) [inline tensorboard/plugins/hparams/api.py]: Did not think too much about this: Is this always desired and is this natural for users?

File: tensorboard/plugins/hparams/api.py [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+      group=None,
+      display_name=None,
+      description=None,
+      dataset_type=None,
+  ):
+    """
+    Args:
+      tag: The tag name of the scalar summary that corresponds to this
+        metric (as a `str`).
+      group: An optional string listing the subdirectory under the
```
Comment (stephanwlee) [inline tensorboard/plugins/hparams/api.py]: Usage question: so I want metric to be written in arbitrary deep path, I should pass a string with the correct path separator for a specific OS?

File: tensorboard/plugins/hparams/hparams_demo.py [hunk 5]
```diff
@@ -261,13 +244,7 @@ def run_all(logdir, verbose=False):
...
   session_index = 0  # across all session groups
   for group_index in xrange(flags.FLAGS.num_session_groups):
-    hparams = {
-        "conv_layers": rng.randint(1, 3),
-        "conv_kernel_size": rng.choice([3, 5]),
-        "dense_layers": rng.randint(1, 3),
-        "dropout": rng.uniform(0.1, 0.4),
-        "optimizer": rng.choice(["adam", "adagrad"])
-    }
+    hparams = {h: sample_uniform(h.domain, rng) for h in HPARAMS}
```
Comment (stephanwlee) [inline tensorboard/plugins/hparams/hparams_demo.py]: Shouldn't it be `{h.name: sample_uniform(h.domain, rng) for h in HPARAMS}`?

 Perhaps the doc on `run` should be updated. Currently, it reads `hparams: A dict mapping hyperparameter names to values.`.

File: tensorboard/plugins/hparams/api.py (hunk 3 above)
Comment (wchargin) [reply tensorboard/plugins/hparams/api.py]: It’s a fair question. Logically, a discrete domain is an unordered set,
so it seemed natural to me that the representation should be invariant
under permutation of `values`. But I could be convinced otherwise.

(Note that the frontend already displays discrete axes as sorted,
irrespective of the domain definition.)
